test/distributed/_composable/fsdp/test_fully_shard_memory.py

```diff
@@ -122,13 +122,13 @@ def _test_fully_shard_training_memory(
         loss.sum().backward()
         mem_mb = self._get_peak_active_memory_mb()
         if reshard_after_forward:
-            # 1x max unsharded block parameters (all-gather), 2x max
+            # 2x max unsharded block parameters (all-gather + copy-out), 2x max
```

Comment:
We have to increase our peak active memory, but this _does not_ increase peak reserved memory, which is what ultimately matters.

The reason is that before, we were implicitly still using memory in the reduce-scatter stream even though the memory was considered freed from the caching allocator's perspective.

Reply:
qq: is memory allocated in non-default stream not accounted by "active_bytes.all.peak"?

Reply:
Memory allocated in non-default streams are counted in `"active_bytes.all.peak"`.

Reply:
Here is my current understanding of what was going on in some more detail:
- We did not keep a ref to the `reduce_scatter_input` (allocated in `reduce_scatter_stream`) after returning from `foreach_reduce`, so it is freed by the caching allocator.
- The subsequent allocations will be for next prefetched all-gather and the next gradient computations, which could define the peak memory. However, at this peak, the `reduce_scatter_input` is considered freed despite a NCCL reduce-scatter kernel using that address.
- This was fine correctness wise because the `dist.reduce_scatter_tensor` is run with `async_op=False`, so the `reduce_scatter_stream` waits on the NCCL reduce-scatter immediately. This means that the next allocation and kernel in the `reduce_scatter_stream` (namely, the next `reduce_scatter_input` and its copy-in) can reuse the current `reduce_scatter_input`'s address. This is the _same_ overlap behavior as we have after this PR, just that we explicitly hold the ref to current `reduce_scatter_input` instead of letting it be considered freed by the allocator while still secretly using it.

This all is kind of nuanced, and I am probably not explaining it too clearly. I think the crux is that (1) because the reduce-scatter uses `async_op=False`, we could free its `reduce_scatter_input` immediately from the caching allocator POV but still be correct since subsequent ops will only run after the reduce-scatter finishes and (2) this happens to coincide with the peak memory point.